### DISCUSSION TITLE
Introduce a way to declaratively control scrolling behavior

### DISCUSSION THREAD

**Discussion by yusufyildirim** [react-native-tvos/react-native-tvos#664] [Feature Requests]
I'll get straight to the point. The default scrolling behavior we have is pretty much useless, to be honest. Here's how it looks:

https://github.com/react-native-tvos/react-native-tvos/assets/22980987/af2004cd-464f-40ad-929a-79b40373d7cb

It doesn't scroll at all until we reach to the end. When it does scroll, it does it in a weird way. In short, it's not suitable for any real application. People disable the scrolling behavior with [scrolEnabled](https://reactnative.dev/docs/scrollview#scrollenabled) prop and implement their own solution by listening `onFocus` events of the children focusables and imperatively calling [scrollToIndex](https://reactnative.dev/docs/virtualizedlist#scrolltoindex) method or things like that.

The good thing about `scrollToIndex` is that it supports `viewOffset` and `viewPosition` to fine-tune the exact behavior you desire. Here's what you can achieve:

https://github.com/react-native-tvos/react-native-tvos/assets/22980987/b2463594-1f73-44a0-a0db-39b6b7ca508c

This is somewhat closer to what you'd expect from a real TV application. It's far from being perfect, though. The amount of effort required by the devs is one aspect of the problem but the other is just the look. It still feels off a bit.

As a solution, I'm proposing we should expose prop(s) from VirtualizedList to make scrolling behavior controllable and handle everything inside to abstract this unnecessary complexity as much as possible.

**Comment by klausbadelt** [comment]
Would be a major relief, and I agree with all of the above.

**Comment by kosso** [comment]
@yusufyildirim Hi there. As I'd mentioned on the other thread [here](https://github.com/react-native-tvos/react-native-tvos/issues/848#issuecomment-3325060435), we need to also think about usage when a FlatList is used as a grid (`numColumns > 1`). 

I have created a sample app to demonstrate the issue with the focus & scrolling on AndroidTV [here](https://github.com/kosso/RNTVReproducer/tree/android-flatlist-grid-focus). 

I have also had a go at trying to control the scrolling myself on a non-scrollable FlatList too. 

And here's a video where I use the sample app to demonstrate and explain the differences  (sound on) ... 

https://github.com/user-attachments/assets/01bf1013-61db-45d8-8b14-0ae1ecf53381

Hope this helps show the sort of things we need. 
Thanks!

ps: Another issue I had to overcome was that in my GridItem component, I attempted to get the layout of the item with the `onLayout`
 event, however the `y` value returned by `event.nativeEvent.layout` is always zero. Regardless of which 'row' it's on.
 
 I couldn't get `measureLayout(relativeToNativeNode, onSuccess, onFail)` to work, so ended up calculating using `measureInWindow(callback)`.

**Reply by kosso** [reply]
I've updated my [sample/experiment](https://github.com/kosso/RNTVReproducer/blob/android-flatlist-grid-focus/RNTVReproducer/screens/GridScreenExperiment.tsx) now with a`scrollEnabled={false}` FlatList as grid which now scrolls & focuses the same way on both platforms - and the same way that tvOS does with a scrollEnabled one, providing space above and below the focused item before reaching the top/bottom edges.

https://github.com/user-attachments/assets/75001b8b-e77e-43b9-998c-fe47d7d704dc

**Reply by kosso** [reply]
Just a little update. At over 250 or so items on a 'grid', AndroidTV was having issues, despite various permutations of `initialNumToRender`, `maxToRenderPerBatch` and `windowSize` etc. 

So, I was reading around something related to 'PureComponent' (after seeing a warning about my large, slow FlatList) and decided to try and wrap the render item component in `memo`. The result was astounding. Now my FlatList grid can navigate up and down through hundreds of items on AndroidTV/FireTV almost as smoothly as AppleTV does.  (I also replaced `Image` with `FastImage` for my thumbnails). 

Another thing I did was detect if a longpress `longUp` or `longDown` event was happening, and disable the `animated` property in the `scrollToOffset`, which very much helped stop the focus getting 'lost'. 

An issue I did come across on AndroidTV too, was if there's a focusable element above the FlatList and you're scrolling up quite quickly, the focus can escape from the FlatList and grab the focusable element above it. To mitigate this, I used a `trapFocusUp` toggle in a TVFocusGuideView which wraps the FlatList, and detected when the focus was on the top row or not.

**Comment by jonybekov** [comment]
> It still feels off a bit.

I kinda understand why it feels off. When we call `scrollToIndex` on `onFocus` event, it first moves focus to the element and then scrolls the view. But in Apple TV, by default, focus and scroll moves together simultaneously which gives very pleasant look. I wasn't able to replicate it with `scrollToIndex` api

**Comment by yusufyildirim** [comment]
@RuudBurger implemented a great solution to this problem here https://github.com/react-native-tvos/react-native-tvos/pull/1068

**Comment by douglowder** [comment]
Closing this discussion, as this feature has been implemented in #1068 .
